// time-ranges/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _timeRange = require('./time-range');

var _timeRange2 = _interopRequireDefault(_timeRange);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var merge = function merge(other) {
  for (var i = 0; i < other.length; ++i) {
    this.add(other.start(i), other.end(i));
  }
};

var findRange = function findRange(value) {
  var i = arguments.length <= 1 || arguments[1] === undefined ? 0 : arguments[1];

  while (i < this._ranges.length && value > this._ranges[i].end) {
    ++i;
  }
  return {
    index: i,
    contained: i < this._ranges.length && value >= this._ranges[i].start
  };
};

var isNumber = function isNumber(num) {
  return typeof num === 'number' && !isNaN(num);
};

var isValidIndex = function isValidIndex(i) {
  return isNumber(i) && i >= 0 && i < this._ranges.length;
};

var TimeRanges = function () {
  function TimeRanges() {
    var other = arguments.length <= 0 || arguments[0] === undefined ? null : arguments[0];

    _classCallCheck(this, TimeRanges);

    this._ranges = [];
    if (other != null) {
      merge.call(this, other);
    }
  }

  _createClass(TimeRanges, [{
    key: 'start',
    value: function start(i) {
      if (!isValidIndex.call(this, i)) {
        throw new Error('Invalid index');
      }
      return this._ranges[i].start;
    }
  }, {
    key: 'end',
    value: function end(i) {
      if (!isValidIndex.call(this, i)) {
        throw new Error('Invalid index');
      }
      return this._ranges[i].end;
    }
  }, {
    key: 'add',
    value: function add(start, end) {
      if (!isNumber(start) || !isNumber(end) || end < start) {
        throw new Error('Invalid range');
      }
      var s = findRange.call(this, start);
      var e = findRange.call(this, end, s.index);
      if (s.contained) {
        start = Math.min(start, this._ranges[s.index].start);
      }
      if (e.contained) {
        end = Math.max(end, this._ranges[e.index].end);
      }
      var len = e.index - s.index + (e.contained ? 1 : 0);
      this._ranges.splice(s.index, len, new _timeRange2.default(start, end));
    }
  }, {
    key: 'length',
    get: function get() {
      return this._ranges.length;
    }
  }]);

  return TimeRanges;
}();

exports.default = TimeRanges;
